csrc/validator_utils.h

```diff
@@ -0,0 +1,349 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <exceptions.h>
+#include <device_lower/utils.h>
+#include <executor_utils.h>
+#include <expr_evaluator.h>
+#include <fusion.h>
+#include <ir/iostream.h>
+
+#include <ATen/cuda/CUDAContext.h>
+
+#include <unordered_map>
+
+namespace nvfuser {
+
+namespace {
+
+struct ValidationConstants {
+  // Tolerances generated from randn + add + sum fusion
+  // compared against double precision
+  std::array<std::array<double, 2>, 20> sum_tolerances_float = {
+      {{4, 1.68222e-06},      {8, 2.23704e-06},      {16, 2.95788e-06},
+       {32, 4.4778e-06},      {64, 6.75395e-06},     {128, 8.57934e-06},
+       {256, 1.30594e-05},    {512, 2.19122e-05},    {1024, 3.3451e-05},
+       {2048, 5.78476e-05},   {4096, 0.000108292},   {8192, 0.00012207},
+       {16384, 0.000136882},  {32768, 0.000248561},  {65536, 0.000407594},
+       {131072, 0.000500901}, {262144, 0.000923019}, {524288, 0.00156909},
+       {1048576, 0.00223107}, {2097152, 0.00343043}}};
```

Comment:
I know this PR didn't introduce this table, but I expected to see a square root formula, and it seems like these do roughly follow a sqrt trend. Modelling rounding error as iid finite-variance noise, the variance of the sum should linearly scale with the number of items being reduced, so for a single sum operation we should use `sqrt(n)` times the epsilon for a given type. If we are doing reduction other than sum it could be more or less: e.g. a product of zero-mean variables should have variance raised to the power `n`, so should use `pow(epsilon, n/2)` for the tolerance.

Reply:
While there is no restriction on the reduction type, the comments indicate the tolerances are for `sum` reductions. We can modify this table and tolerance computations (preferably in a separate PR though), but it seems that the values have been working fine for the existing tests. What do you think?